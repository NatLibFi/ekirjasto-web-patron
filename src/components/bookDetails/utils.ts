import { BookData } from "opds-web-client/lib/interfaces";
import * as moment from "moment";

export const getAvailabilityString = (book: BookData): string => {
  if (isOpenAccess(book)) {
    return "This open-access book is available to keep.";
  }

  if (isBorrowed(book)) {
    const availableUntil = book?.availability?.until;
    if (availableUntil) {
      const timeLeft = moment(availableUntil).fromNow(true);
      return `You have this book on loan for ${timeLeft}`;
    }
    // you have borrowed the book but it is unknown until when
    return "You have this book on loan";
  }

  const availableCopies = book?.copies?.available;
  const totalCopies = book?.copies?.total;
  const totalHolds = book?.holds?.total;
  const holdsPosition = book?.holds?.position;

  let availabilityString = "";
  // there is a number of copies available that is known
  if (availableCopies && totalCopies) {
    // show how many available, even if all are taken
    availabilityString += `${availableCopies} of ${totalCopies} copies available.`;

    // there is a queue
    if (totalHolds && availableCopies === 0) {
      availabilityString += ` ${totalHolds} patrons in queue.`;

      if (isReserved(book) && typeof holdsPosition === "number") {
        availabilityString += ` Your hold position: ${holdsPosition}.`;
      }
    }

    return availabilityString;
  }
  // the number of available copies is unknown
  return "Availability unknown";
};

export const isReserved = (book: BookData): boolean => {
  return book?.availability?.status === "reserved";
};

export const isOpenAccess = (book: BookData): boolean => {
  const numLinks = book?.openAccessLinks?.length ?? 0;
  return numLinks > 0;
};

export const isBorrowed = (book: BookData): boolean => {
  const numberOfLinks = book?.fulfillmentLinks?.length ?? 0;
  return numberOfLinks > 0;
};

type CirculationLink = {
  isOpenAccess: boolean;
  isBorrowed: boolean;
};

export const getCirculationLinkMap = (book: BookData) => {
  /**
   * This should mirror the OPDS logic, but instead of returning an array
   * of components, it returns a map of links, which lets the consumer render
   * those links however they desire. It will give you all the possible links,
   * and the info needed to render them properly
   * - Open Access links are the first option, shown whether borrowed or not
   * - If not open access, you must borrow first. set isBorrowed to false
   * -
   */
};
