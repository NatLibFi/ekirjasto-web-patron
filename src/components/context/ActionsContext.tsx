import * as React from "react";
import ActionsCreator from "opds-web-client/lib/actions";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";
import useThunkDispatch from "../../hooks/useThunkDispatch";

const ActionsContext = React.createContext<ActionsCreator | undefined>(
  undefined
);

export const ActionsProvider: React.FC<{ proxyUrl?: string }> = ({
  children,
  proxyUrl
}) => {
  // create our datafetcher
  const fetcher = new DataFetcher({
    proxyUrl,
    adapter
  });
  const actions = new ActionsCreator(fetcher);

  return (
    <ActionsContext.Provider value={actions}>
      {children}
    </ActionsContext.Provider>
  );
};

export function useActions() {
  const context = React.useContext(ActionsContext);
  const dispatch = useThunkDispatch();
  if (typeof context === "undefined") {
    throw new Error("useActions must be used within a ActionsProvider");
  }
  return { actions: context, dispatch };
}

export default ActionsContext;
