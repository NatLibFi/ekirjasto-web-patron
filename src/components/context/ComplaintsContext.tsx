import * as React from "react";
import complaintsReducer, {
  initState,
  ComplaintsState
} from "../../reducers/complaints";
import useThunkReducer from "../../hooks/useThunkReducer";

export const ComplaintsStateContext = React.createContext<
  ComplaintsState | undefined
>(undefined);
export const ComplaintsDispatchContext = React.createContext<
  React.Dispatch<any> | undefined
>(undefined);

export const ComplaintsContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useThunkReducer(complaintsReducer, initState);

  return (
    <ComplaintsStateContext.Provider value={state}>
      <ComplaintsDispatchContext.Provider value={dispatch}>
        {children}
      </ComplaintsDispatchContext.Provider>
    </ComplaintsStateContext.Provider>
  );
};

function useComplaintsState() {
  const state = React.useContext(ComplaintsStateContext);
  const dispatch = React.useContext(ComplaintsDispatchContext);
  // const { actions, dispatch: _unusedOpdsDispatch } = useActi/ons();

  if (typeof state === "undefined" || typeof dispatch === "undefined") {
    throw new Error(
      "useComplaintsState must be used within a ComplaintsContextProvider"
    );
  }
  return {
    complaintsState: state,
    complaintsDispatch: dispatch
    // recommendationsActions: actions
  };
}

export default useComplaintsState;
