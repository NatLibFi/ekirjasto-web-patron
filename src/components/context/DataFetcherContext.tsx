import * as React from "react";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";

const DataFetcherContext = React.createContext<DataFetcher | undefined>(
  undefined
);

export function DataFetcherProvider({ children, proxyUrl = null }) {
  // create our datafetcher
  const fetcher = new DataFetcher({
    proxyUrl,
    adapter
  });

  return (
    <DataFetcherContext.Provider value={fetcher}>
      {children}
    </DataFetcherContext.Provider>
  );
}

export function useDataFetcher() {
  const context = React.useContext(DataFetcherContext);
  if (typeof context === "undefined") {
    throw new Error("useDataFetcher must be used within a DataFetcherProvider");
  }
  return context;
}

export default DataFetcherContext;
