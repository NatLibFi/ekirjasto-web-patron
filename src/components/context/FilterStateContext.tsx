/** @jsx jsx */
import { jsx } from "theme-ui";
import * as React from "react";

type FilterState = "ALL" | "EBOOKS" | "AUDIOBOOKS";
type FilterContext = {
  filterState: FilterState;
  setFilter: React.Dispatch<React.SetStateAction<FilterState>>;
};

const FilterStateContext = React.createContext<FilterContext | undefined>(
  undefined
);

export const FilterStateProvider: React.FC = ({ children }) => {
  const [filterState, setFilter] = React.useState<FilterState>("ALL");

  const value = React.useMemo(() => ({ filterState, setFilter }), [
    filterState,
    setFilter
  ]);

  return (
    <FilterStateContext.Provider value={value}>
      {children}
    </FilterStateContext.Provider>
  );
};

export default function useFilterState() {
  const context = React.useContext(FilterStateContext);
  if (typeof context === "undefined") {
    throw new Error("useFilterState must be used within a FilterStateProvider");
  }
  return context;
}
