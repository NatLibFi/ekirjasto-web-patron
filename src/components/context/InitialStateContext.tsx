import * as React from "react";
import { State as CatalogState } from "opds-web-client/lib/state";

const InitialStateContext = React.createContext<CatalogState>(null);

export const InitialStateProvider: React.FC<{ initialState: CatalogState }> = ({
  initialState,
  children
}) => (
  <InitialStateContext.Provider value={initialState}>
    {children}
  </InitialStateContext.Provider>
);

export default InitialStateContext;
