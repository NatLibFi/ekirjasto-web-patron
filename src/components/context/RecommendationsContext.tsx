import * as React from "react";
import collection, {
  initialState
} from "opds-web-client/lib/reducers/collection";
import useThunkReducer from "../../hooks/useThunkReducer";

export const RecommendationsStateContext = React.createContext(null);
export const RecommendationsDispatchContext = React.createContext(null);

export const RecommendationsContextProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useThunkReducer(collection, initialState);

  return (
    <RecommendationsStateContext.Provider value={state}>
      <RecommendationsDispatchContext.Provider value={dispatch}>
        {children}
      </RecommendationsDispatchContext.Provider>
    </RecommendationsStateContext.Provider>
  );
};

const useRecommendationsState = () => {
  const state = React.useContext(RecommendationsStateContext);
  const dispatch = React.useContext(RecommendationsDispatchContext);
  // possibly bind all the actions with dispatch here?
  return [state, dispatch];
};

export default useRecommendationsState;
