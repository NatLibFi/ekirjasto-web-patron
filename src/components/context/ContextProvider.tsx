import * as React from "react";
import { PathFor, PreloadedData } from "interfaces";
import UrlShortener from "UrlShortener";
import { LibraryProvider } from "./LibraryContext";
import PathForProvider from "opds-web-client/lib/components/context/PathForContext";
import { RouterProvider } from "./RouterContext";
import OPDSStore from "opds-web-client/lib/components/context/StoreContext";
import { RecommendationsProvider } from "./RecommendationsContext";
import { ActionsProvider } from "opds-web-client/lib/components/context/ActionsContext";
import { Provider as ReakitProvider } from "reakit";
import { State } from "opds-web-client/lib/state";
import { Store } from "redux";
import DataFetcher from "opds-web-client/lib/DataFetcher";
import ActionsCreator from "opds-web-client/lib/actions";
import { adapter } from "opds-web-client/lib/OPDSDataAdapter";
import basicAuthPlugin from "auth/basicAuthPlugin";
import samlAuthPlugin from "auth/samlAuthPlugin";
import getPathFor from "utils/getPathFor";
import { LinkUtilsProvider } from "./LinkUtilsContext";

type ProviderProps = PreloadedData & {
  // we allow custom store and actions
  // to be passed in for the sake of mocking during testing
  store?: Store<State>;
  actions?: ActionsCreator;
  fetcher?: DataFetcher;
};
/**
 * Combines all of the apps context provider into a single component for simplicity
 */
const AppContextProvider: React.FC<ProviderProps> = ({
  children,
  library,
  shortenUrls,
  initialState,
  store,
  actions,
  fetcher
}) => {
  const libraryId = library.id;
  const urlShortener = new UrlShortener(library.catalogUrl, shortenUrls);
  const pathFor: PathFor = getPathFor(urlShortener, libraryId);
  const computedFetcher = fetcher ?? new DataFetcher({ adapter });
  const computedActions = actions ?? new ActionsCreator(computedFetcher);

  return (
    <ReakitProvider>
      <RouterProvider>
        <PathForProvider pathFor={pathFor}>
          <OPDSStore
            initialState={initialState}
            store={store}
            authPlugins={[basicAuthPlugin, samlAuthPlugin]}
          >
            <RecommendationsProvider>
              <ActionsProvider
                actions={computedActions}
                fetcher={computedFetcher}
              >
                <LibraryProvider library={library}>
                  <LinkUtilsProvider
                    library={library}
                    urlShortener={urlShortener}
                  >
                    {children}
                  </LinkUtilsProvider>
                </LibraryProvider>
              </ActionsProvider>
            </RecommendationsProvider>
          </OPDSStore>
        </PathForProvider>
      </RouterProvider>
    </ReakitProvider>
  );
};

export default AppContextProvider;
