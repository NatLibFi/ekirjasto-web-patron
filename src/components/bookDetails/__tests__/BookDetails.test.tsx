import * as React from "react";
import { render, fixtures, actions, waitFor } from "../../../test-utils";
import merge from "deepmerge";
import { BookDetails } from "../index";
import { State } from "owc/state";
import { BookData, CollectionData } from "owc/interfaces";
import * as complaintActions from "../../../hooks/useComplaints/actions";
import { RecommendationsStateContext } from "../../context/RecommendationsContext";
import userEvent from "@testing-library/user-event";
import ReportProblem from "../ReportProblem";
import * as env from "../../../utils/env";

const mockSetCollectionAndBook = jest.fn().mockResolvedValue({});

/**
 * We need to mock actions.fetchCollection so that
 * ReportProblem doesn't actually fetch anything
 */
jest
  .spyOn(actions, "fetchCollection")
  .mockImplementation(() => () => Promise.resolve({} as CollectionData));

const makeStateWithBook = (book: BookData = fixtures.book): State =>
  merge<State>(fixtures.initialState, {
    book: {
      isFetching: false,
      url: "/book",
      error: null,
      data: book
    }
  });

describe("book details page", () => {
  test("shows loading state", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: merge<State>(fixtures.initialState, {
          book: {
            isFetching: true,
            url: "/book",
            error: null,
            data: null
          }
        })
      }
    );
    expect(
      utils.getByRole("heading", { name: "Loading..." })
    ).toBeInTheDocument();
  });

  test("handles error state", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: merge<State>(fixtures.initialState, {
          book: {
            isFetching: false,
            url: "/book",
            error: {
              response: "Some error",
              status: 401,
              url: "/book-error-url"
            },
            data: null
          }
        })
      }
    );
    /**
     * We will snapshot test this component because it isn't complicated and should prettymuch remain the same.
     */
    expect(utils.container).toMatchSnapshot();
  });

  test("shows categories", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    const categories = fixtures.book.categories?.join(", ") as string;
    expect(utils.getByText(categories));
    expect(utils.getByText("Categories:"));
  });

  test("doesn't show categories when there aren't any", () => {
    const bookWithoutCategories: BookData = merge(fixtures.book, {
      categories: null
    });
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithoutCategories)
      }
    );
    expect(utils.queryByText("Categories:")).toBeFalsy();
  });

  test("shows publisher", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    const publisher = fixtures.book.publisher as string;

    expect(utils.getByText(publisher)).toBeInTheDocument();
    expect(utils.getByText("Publisher:")).toBeInTheDocument();
  });

  test("shows fulfillment card in open-access state", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    expect(
      utils.getByText("This open-access book is available to keep forever.")
    );
    expect(
      utils.getByRole("button", { name: "Borrow to read on a mobile device" })
    ).toBeInTheDocument();
  });

  test("does not show simplyE callout when NEXT_PUBLIC_COMPANION_APP is 'openebooks'", () => {
    (env.NEXT_PUBLIC_COMPANION_APP as string) = "openebooks";
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    expect(
      utils.queryByText("Read Now. Read Everywhere.")
    ).not.toBeInTheDocument();

    expect(utils.queryByText("SimplyE Logo")).not.toBeInTheDocument();
    expect(
      utils.queryByText(
        "Browse and read our collection of eBooks and Audiobooks right from your phone."
      )
    ).not.toBeInTheDocument();
  });

  test("shows simplyE callout when NEXT_PUBLIC_COMPANION_APP is 'simplye'", async () => {
    (env.NEXT_PUBLIC_COMPANION_APP as string) = "simplye";
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    expect(utils.getByText("Read Now. Read Everywhere.")).toBeInTheDocument();
    expect(utils.getByLabelText("SimplyE Logo")).toBeInTheDocument();
    expect(
      utils.getByText(
        "Browse and read our collection of eBooks and Audiobooks right from your phone."
      )
    ).toBeInTheDocument();

    const iosBadge = utils.getByRole("link", {
      name: "Download SimplyE on the Apple App Store",
      hidden: true // it is initially hidden by a media query, only displayed on desktop
    });
    expect(iosBadge).toBeInTheDocument();
    expect(iosBadge).toHaveAttribute(
      "href",
      "https://apps.apple.com/us/app/simplye/id1046583900"
    );

    const googleBadge = utils.getByRole("link", {
      name: "Get SimplyE on the Google Play Store",
      hidden: true // hidden initially on mobile
    });
    expect(googleBadge).toBeInTheDocument();
    expect(googleBadge).toHaveAttribute(
      "href",
      "https://play.google.com/store/apps/details?id=org.nypl.simplified.simplye&pcampaignid=pcampaignidMKT-Other-global-all-co-prtnr-py-PartBadge-Mar2515-1"
    );
  });

  test("shows recommendation lanes", () => {
    const utils = render(
      <RecommendationsStateContext.Provider
        value={{
          ...fixtures.recommendationsState
        }}
      >
        <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />
      </RecommendationsStateContext.Provider>,
      {
        initialState: makeStateWithBook()
      }
    );
    expect(utils.getByText("Recommendations")).toBeInTheDocument();
    expect(
      utils.getByRole("heading", { name: "Jane Austen" })
    ).toBeInTheDocument();
  });
});

/**
 * mock fetchComplaintTypes and postComplaint
 */
const mockBoundFetchComplaintTypes = jest
  .fn()
  .mockResolvedValue(["complaint-type"]);
const fetchComplaintTypesSpy = jest
  .spyOn(complaintActions, "fetchComplaintTypes")
  .mockReturnValue(mockBoundFetchComplaintTypes);
const postComplaintSpy = jest.spyOn(complaintActions, "postComplaint");

const bookWithReportUrl = merge<BookData>(fixtures.book, {
  raw: {
    link: [
      {
        $: {
          rel: {
            value: "issues"
          },
          href: {
            value: "/report-url"
          }
        }
      }
    ]
  }
});

describe("report problem", () => {
  test("shows report problem link", () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );

    const reportProblemLink = utils.getByTestId("report-problem-link");
    expect(reportProblemLink).toBeInTheDocument();
  });

  test("shows form (only) when clicked", async () => {
    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook()
      }
    );
    // make sure it's not visible at first
    expect(utils.getByLabelText("Complaint Type")).not.toBeVisible();

    const reportProblemLink = utils.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // one for the original report problem button, one for the form
    expect(utils.getAllByText("Report a problem")).toHaveLength(2);
    expect(utils.getByLabelText("Complaint Type")).toBeVisible();
    expect(utils.getByLabelText("Details")).toBeVisible();
    expect(utils.getByText("Cancel")).toBeVisible();
    expect(utils.getByText("Submit")).toBeVisible();
    expect(utils.getByText("Submit")).toHaveAttribute("type", "submit");
  });

  test("fetches complaint types", async () => {
    render(<BookDetails setCollectionAndBook={mockSetCollectionAndBook} />, {
      initialState: makeStateWithBook(bookWithReportUrl)
    });
    expect(mockBoundFetchComplaintTypes).toHaveBeenCalledTimes(1);
    expect(mockBoundFetchComplaintTypes).toHaveBeenCalledWith("/report-url");
  });

  test("submits", async () => {
    /**
     * Mock the fetchComplaintTypes function so that it actually
     * loads the complaintTypes into the state, so we can then
     * select them. I tried to mock the state value itself, but that
     * turned out to be quite difficult.
     */
    const complaintTypes = ["complaint-type-a", "complaint-type-b"];
    fetchComplaintTypesSpy.mockImplementationOnce(dispatch => _url => {
      dispatch({
        type: "FETCH_COMPLAINT_TYPES_SUCCESS",
        types: complaintTypes
      });
      return Promise.resolve(complaintTypes);
    });

    const mockBoundPostComplaint = jest.fn().mockResolvedValue(["some-string"]);
    postComplaintSpy.mockReturnValue(_url => mockBoundPostComplaint);

    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithReportUrl)
      }
    );
    // open the form
    const reportProblemLink = utils.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // fill the form
    userEvent.selectOptions(
      utils.getByLabelText("Complaint Type"),
      "complaint-type-b"
    );
    userEvent.type(utils.getByLabelText("Details"), "Some issue happened.");
    // submit the form
    userEvent.click(utils.getByText("Submit"));

    // actually posted the complaint
    await waitFor(() =>
      expect(mockBoundPostComplaint).toHaveBeenCalledTimes(1)
    );
    expect(mockBoundPostComplaint).toHaveBeenCalledWith({
      type: "complaint-type-b",
      detail: "Some issue happened."
    });
  });

  test("displays thank you after submitting", async () => {
    const complaintTypes = ["complaint-type-a", "complaint-type-b"];
    fetchComplaintTypesSpy.mockImplementationOnce(dispatch => _url => {
      dispatch({
        type: "FETCH_COMPLAINT_TYPES_SUCCESS",
        types: complaintTypes
      });
      return Promise.resolve(complaintTypes);
    });

    // make postComplaint dispatch a success message
    postComplaintSpy.mockImplementation(dispatch => _url => _data => {
      dispatch({ type: "POST_COMPLAINT_SUCCESS" });
      return Promise.resolve(["some string"]);
    });

    const utils = render(
      <BookDetails setCollectionAndBook={mockSetCollectionAndBook} />,
      {
        initialState: makeStateWithBook(bookWithReportUrl)
      }
    );
    // open the form
    const reportProblemLink = utils.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // fill the form
    userEvent.selectOptions(
      utils.getByLabelText("Complaint Type"),
      "complaint-type-b"
    );
    userEvent.type(utils.getByLabelText("Details"), "Some issue happened.");
    // submit the form
    userEvent.click(utils.getByText("Submit"));

    // shows thank you message
    expect(
      await utils.findByText("Your problem was reported. Thank you!")
    ).toBeInTheDocument();
    expect(await utils.findByText("Done")).toBeInTheDocument();

    // can close the form
    userEvent.click(utils.getByText("Done"));
    expect(utils.getByText("Complaint Type")).not.toBeVisible();
  });

  test("displays client error when unfilled", async () => {
    const mockBoundPostComplaint = jest.fn().mockResolvedValue(["some-string"]);
    postComplaintSpy.mockReturnValue(_url => mockBoundPostComplaint);

    const utils = render(<ReportProblem book={fixtures.book} />, {
      initialState: makeStateWithBook(bookWithReportUrl)
    });
    // open the form
    const reportProblemLink = utils.getByTestId("report-problem-link");
    userEvent.click(reportProblemLink);

    // don't fill the form
    // submit the form
    userEvent.click(utils.getByText("Submit"));

    expect(
      await utils.findByText("Error: Please choose a type")
    ).toBeInTheDocument();
    expect(
      await utils.findByText("Error: Please enter details about the problem.")
    ).toBeInTheDocument();

    expect(mockBoundPostComplaint).toHaveBeenCalledTimes(0);
  });
});
