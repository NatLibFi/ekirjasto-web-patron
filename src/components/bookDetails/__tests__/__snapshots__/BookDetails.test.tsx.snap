// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book details page handles error state 1`] = `
.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  max-width: 70%;
}

.emotion-0 {
  font-weight: 700;
}

.emotion-2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: -apple-system,BlinkMacSystemFont,system-ui,sans-serif;
  font-size: 0.875rem;
  font-weight: 300;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  padding-left: 16px;
  padding-right: 16px;
  border: 0;
  border-radius: 2px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: var(--theme-ui-colors-brand-primary,#337ab7);
  height: 32px;
  min-width: 32px;
  color: white;
  fill: white;
  margin-top: 16px;
}

.emotion-2:hover {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-2:focus,
.emotion-2:hover {
  background-color: #2d6da3;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-2:focus {
  box-shadow: 0 0 0 0.2em rgba(0,109,255,0.4);
}

.emotion-2:active {
  background-color: #285f8f;
}

.emotion-2:disabled {
  background-color: var(--theme-ui-colors-ui-gray-light,#e0e0e0);
  color: var(--theme-ui-colors-ui-gray-extraDark,#424242);
  cursor: default;
}

<div>
  <section
    aria-label="Book details"
    class="emotion-4"
  >
    <div
      class="emotion-3"
    >
      <p>
        There was a problem fetching this book. Please refresh the page or return home.
      </p>
      <div>
        <span
          class="emotion-0"
        >
          Error Code: 
        </span>
        401
      </div>
      <div>
        <span
          class="emotion-0"
        >
          Error Message: 
        </span>
        Some error
      </div>
      <a
        class="emotion-2"
        href="/"
      >
        Return Home
      </a>
    </div>
  </section>
</div>
`;
